starport/services/networkbuilder/simulation.go

```diff
@@ -0,0 +1,219 @@
+package networkbuilder
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"context"
+	"github.com/cenkalti/backoff"
+	"github.com/otiai10/copy"
+	"github.com/pelletier/go-toml"
+	"github.com/tendermint/starport/starport/pkg/availableport"
+	"github.com/tendermint/starport/starport/pkg/cmdrunner"
+	"github.com/tendermint/starport/starport/pkg/cmdrunner/step"
+	"github.com/tendermint/starport/starport/pkg/cosmosver"
+	"github.com/tendermint/starport/starport/pkg/gomodulepath"
+	"github.com/tendermint/starport/starport/pkg/httpstatuschecker"
+	"github.com/tendermint/starport/starport/pkg/xurl"
+	"github.com/tendermint/starport/starport/services/chain"
+	"io"
+	"io/ioutil"
+	"net/url"
+	"os"
+	"path"
+	"path/filepath"
+	"strings"
+	"time"
+)
+
+const ErrValidatorSetNil = "validator set is nil in genesis and still empty after InitChain"
+
+// VerifyProposals generates a genesis file from the current launch information and proposals to verify
+// The function returns false if the generated genesis is invalid
+func (b *Builder) VerifyProposals(ctx context.Context, chainID string, proposals []int, commandOut io.Writer) (bool, error) {
+	chainInfo, err := b.ShowChain(ctx, chainID)
+	if err != nil {
+		return false, err
+	}
+
+	// Get the simulated launch information from these proposals
+	simulatedLaunchInfo, err := b.SimulatedLaunchInformation(ctx, chainID, proposals)
+	if err != nil {
+		return false, err
+	}
+
+	// find out the app's name form url
+	u, err := url.Parse(chainInfo.URL)
+	if err != nil {
+		return false, err
+	}
+	importPath := path.Join(u.Host, u.Path)
+	path, err := gomodulepath.Parse(importPath)
+	if err != nil {
+		return false, err
+	}
+	app := chain.App{
+		ChainID: chainID,
+		Name:    path.Root,
+		Version: cosmosver.Stargate,
+	}
+	chainCmd, err := chain.New(app, true, chain.LogSilent)
+	if err != nil {
+		return false, err
+	}
+
+	// get app dir
+	homedir, err := os.UserHomeDir()
+	if err != nil {
+		return false, err
+	}
+	appHome := filepath.Join(homedir, app.ND())
+
+	// create a temporary dir that holds the genesis to test
+	tmpHome, err := ioutil.TempDir("", app.ND() + "*")
+	if err != nil {
+		return false, err
+	}
+	defer os.RemoveAll(tmpHome)
+	err = copy.Copy(appHome, tmpHome)
+	if err != nil {
+		return false, err
+	}
+
+	// generate the genesis to test
+	if err := generateGenesis(ctx, tmpHome, chainInfo, simulatedLaunchInfo, chainCmd); err != nil {
+		fmt.Fprintf(commandOut, "error generating the genesis: %s\n", err.Error())
+		return false, nil
+	}
+
+	// set the config with random ports to test the start command
+	addressAPI, err := setSimulationConfig(tmpHome)
+	if err != nil {
+		return false, err
+	}
+
+	// run validate-genesis command on the generated genesis
+	err = cmdrunner.New().Run(ctx, step.New(
+		step.Exec(
+			app.D(),
+			"validate-genesis",
+			"--home",
+			tmpHome,
+		),
+		step.Stderr(commandOut),	// This is the error of the verifying command, therefore this is the same as stdout
```

Comment:
Instead of piping this error to the external callers, we can make `VerifyProposals` to return a plain `error` value. Similar to the error handling here: https://github.com/tendermint/starport/blob/49aa840e0df1bdefaa45a0cb57c0a725ce5f4aff/starport/services/chain/init.go#L278

Reply:
`VerifyProposals` should not return an error because the function didn't fail. The function succeeded and the proposals are not valid

Reply:
In this case, can we assume that even if some data comes from the stderr, the command we execute does exit with error code `1`?

Reply:
From what I have seen, `validate-genesis` exit with one if the genesis is not valid otherwise 0